File: code/abc/010d.cpp
#include <map>
#include <set>
#include <vector>
#include <algorithm>
#include <iostream>
#include <bitset>
#include <cassert>
#include <queue>
#include <random>
#include <stack>
#include <iomanip>

using namespace std;
template<class T> inline bool chmax(T& a, T b) { if (a < b) { a = b; return true; } return false; }
template<class T> inline bool chmin(T& a, T b) { if (a > b) { a = b; return true; } return false; }

typedef unsigned int uint;
typedef long long llong;
typedef unsigned long long ullong;
typedef long double ldouble;

typedef vector<llong> vecllong;
typedef vector<vecllong> vvecllong;

const llong MOD = 1e9 + 7;
const llong INF = 1e17;

#define FOR(i, n) for (llong i = 0; i < n; i++)
#define FORS(i, a, b) for (llong i = a; i < b; i++)
#define FORR(i, n) for (llong i = n; i > 0; i++)

struct edge {
    llong capacity;
    llong to;
    llong reverseEdge;
};

class Dinic {
    private:
        llong size;
        vector<vector<edge>> edges;
        vecllong depth;
        vecllong progress;

        void bfs(llong s) {
            vecllong depth(this->size, -1);
            depth[s] = 0;
            deque<llong> que(1, s);

            while (!que.empty()) {
                llong now = que.front();
                que.pop_front();
                for_each(this->edges[now].begin(), this->edges[now].end(), [&](edge e) {
                    if (e.capacity > 0 && depth[e.to] < 0) {
                        depth[e.to] = depth[now] + 1;
                        que.push_back(e.to);
                    }
                });
            }
            this->depth = depth;
        }

        llong dfs(llong s, llong t, llong flow) {
            if (s == t) return flow;
            vector<edge> es = this->edges[s];
            FORS(i, this->progress[s], es.size()) {
                this->progress[s] = i;
                edge e = es[i];
                if (e.capacity == 0 || this->depth[s] >= this->depth[e.to]) continue;

                llong d = this->dfs(e.to, t, min(flow, e.capacity));

                if (d == 0) continue;

                this->edges[s][i].capacity -= d;
                this->edges[e.to][e.reverseEdge].capacity += d;

                return d;
            }
            return 0;
        };

    public:
        Dinic (llong size) {
            this->init(size);
        };

        void init(llong size) {
            this->size = size;
            this->edges = vector<vector<edge>>(size, vector<edge>(0));
        };

        void addEdge(llong fr, llong to, llong cap) {
            this->edges[fr].push_back({cap, to, this->edges[to].size()});
            this->edges[to].push_back({cap, fr, this->edges[fr].size() - 1});
        };

        llong maxFlow(llong s, llong t) {
            llong flow = 0;
            while (true) {
                this->bfs(s);
                if (this->depth[t] < 0) return flow;
                this->progress = vecllong(this->size, 0);
                llong currentFlow = this->dfs(s, t, INF);
                while (currentFlow > 0) {
                    flow += currentFlow;
                    currentFlow = this->dfs(s, t, INF);
                }
            }
        }
};

int main(void) {
    cin.tie(0);
    ios::sync_with_stdio(false);

    llong N, G, E;
    cin >> N >> G >> E;

    vecllong P(G);
    FOR(i, G) cin >> P[i];

    vvecllong edges(N, vecllong(0));
    vector<pair<llong, llong>> A(E);

    FOR(i, E) {
        llong fr, to;
        cin >> fr >> to;
        edges[fr].push_back(to);
        edges[to].push_back(fr);
        A[i] = {fr, to};
    }

    vecllong depth(N + 1, -1);
    stack<llong> st;
    st.push(0);
    depth[0] = 0;

    while (!st.empty()) {
        llong now = st.top();
        st.pop();
        for_each(edges[now].begin(), edges[now].end(), [&](llong to) {
            if (depth[to] == -1) {
                depth[to] = depth[now] + 1;
                st.push(to);
            }
        });
    }

    Dinic flow(N + 1);
    FOR(i, E) {
        llong fr, to;
        fr = A[i].first;
        to = A[i].second;
        if (depth[fr] > depth[to]) swap(fr, to);
        flow.addEdge(fr, to, 1);
    }

    FOR(i, G) flow.addEdge(P[i], N, 1);

    cout << flow.maxFlow(0, N) << endl;

    return 0;
}

File: AtCoder/abc/010d.py
import sys
import heapq
from operator import itemgetter
from collections import deque, defaultdict, Counter
from bisect import bisect_left, bisect_right
input = sys.stdin.readline
sys.setrecursionlimit(10 ** 7)
MOD = 10**9 + 7
INF = float('inf')

class Dinic:
    def __init__(self, size):
        self.size = size
        self.edges = [[] for _ in range(size)]
        self.depth = None
        self.progress = None

    def addEdge(self, fr, to, cap):
        self.edges[fr].append([cap, to, len(self.edges[to])])
        self.edges[to].append([0, fr, len(self.edges[fr]) - 1])

    def bfs(self, s):
        depth = [-1] * self.size
        depth[s] = 0
        que = deque([s])

        while que:
            now = que.popleft()
            for cap, to, _ in self.edges[now]:
                if cap > 0 and depth[to] < 0:
                    depth[to] = depth[now] + 1
                    que.append(to)
        self.depth = depth

    def dfs(self, s, t, flow):
        if s == t:
            return flow

        for i, (cap, to, rev) in enumerate(self.edges[s][self.progress[s]:], start=self.progress[s]):
            self.progress[s] = i
            if cap == 0 or self.depth[s] >= self.depth[to]:
                continue
            d = self.dfs(to, t, min(flow, cap))
            if d == 0:
                continue
            self.edges[s][i][0] -= d
            self.edges[to][rev][0] += d
            return d
        return 0

    def maxFlow(self, s, t):
        flow = 0
        while True:
            self.bfs(s)
            if self.depth[t] < 0:
                return flow
            self.progress = [0] * self.size
            currentFlow = self.dfs(s, t, float('inf'))
            while currentFlow > 0:
                flow += currentFlow
                currentFlow = self.dfs(s, t, float('inf'))

def sol():
    N, _, E = map(int, input().split())
    P = list(map(int, input().split()))
    A = []
    edges = [[] for _ in range(N)]

    for _ in range(E):
        fr, to = map(int, input().split())
        A.append((fr, to))
        edges[fr].append(to)
        edges[to].append(fr)

    depth = [-1] * (N + 1)
    st = deque([0])
    depth[0] = 0
    while st:
        now = st.popleft()
        for to in edges[now]:
            if depth[to] == -1:
                depth[to] = depth[now] + 1
                st.append(to)

    flow = Dinic(N + 1)
    for fr, to in A:
        if depth[fr] > depth[to]:
            fr, to = to, fr
        flow.addEdge(fr, to, 1)

    for p in P:
        flow.addEdge(p, N, 1)

    print(flow.maxFlow(0, N))

sol()